text/0000-numeric-literal-types.md

```diff
@@ -0,0 +1,318 @@
+context, they will attempt to coerce into a sized integer type, defaulting
+to `i32` or `f64` if there isn't an obvious choice. This occurs silently,
+since one almost always wants a sized integer:
+```rust
+let x = 42; // 42 types as ulit, but since a let binding requires a 
+            // Sized value, it tries to coerce to a sized integer. since
+            // there isn't an obvious one, it picks i32. Hence, x: i32.
+
+let y = 42u32; // 42u32 has type u32, so no coersion occurs.
+let z: u32 = 42; // ulit coerces to u32, since it's the required type
+```
+
+Literal types are otherwise *mostly* like normal integers. They support
+arithmetic and comparisons (but don't implement any `std::ops` traits, since
+they're not `Sized`). Like any DST, they can be passed around behind references.
+You can even write
+```rust
+const REALLY_BIG: &ulit = &1000000000000000000000;
+// analogous to
+const HELLO_WORLD: &str = "Hello, world!";
+```
+You can then use `REALLY_BIG` anywhere you'd use the literal, instead. The
+reference types `&'static ulit` and `&'static flit` will automatically coerce 
+into any numeric type, via dereferencing.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC introduces two new DSTs: `ulit` and `flit`. Note that we do not
+introduce `ilit`; this is because it is not possible to write down a literal
+for a negative number, since `-1` is *really* `1.neg()`. These types behave
+like most DSTs, with a few exceptions:
+
+If either type is used in a context that requires a `Sized` type 
+(a function call, a let binding, a generic paramter, etc), they will
+coerce according to the current typing rules for literals: whatever
+is infered as correct, or `i32`/`f64` as a fallback. Note that `as` casts
+do what is expected. See [unresolved-questions] for alternative
+ways we could perform the coersions.
+
+For ergonomic reasons,
+static references to either type are dereferenced automatically in `Sized`
+context. This is to support the following pattern:
+```rust
+const FOO: &ulit = 0b0101_0101;
+
+let x: u8 = 5;
+let y = 5 & FOO; // here `FOO` is coerced from `&ulit` to `u8`
+```
+
+The representation of `ulit` and `flit` is unspecified, but this RFC suggests
+representations. Note that the compiler is *not* required to use these;
+they are merely a suggestion for what a good representation would be.
+```rust
+// represented as an array of bytes in the target endianness;
+// this endianness choice means a coersion is just a memcpy
+struct ulit([u8]);
+
+// represented as a ratio of target-endian, arbitrary-length
+// integers. this is chosen over an IEEE-like base-2 notation,
+// which would require rounding. unfortunately, this requires
+// an fdiv for coersion, though this is not a problem, since
+// it is unlikely that an flit will need to be coerced at runtime
+struct flit {
+    middle: usize,
+    bytes: [u8]
```

Comment:
This representation doesn't work either, at least not the "fdiv to convert" bit, since to do that you first need to convert numerator and denominator to the float format.

The "ratio of bignums" aspect is at least lossless and can therefore be salvaged by a more expensive conversion algorithms, but it's not *more* useful than other representations for those algorithms. Frankly, I don't think that there's anything better than a plain old string for *representing* literals.

In any case, the conversion will be rather expensive when done at runtime. In the worst case it requires at least one >1000 bit integer division+remainder calculations, to the best of my knowledge even multiple. Yet another reason to not permit such values to leak into runtime.

Reply:
@kennytm I don't quite understand what you mean-- the `middle` field indicated where the numerator ends and the denominator starts? I wrote `([u8], [u8])` originally but I didn't want to have to explain the pseudocode.

Also... this seems like a problem, since neither a ratio or IEEE representation seem better... the former makes scientific notation awful, and the latter makes most decimals require truncation. I think @rkruppe has a point- I don't intend these values to ever reach a runtime context, so in practice a string representation with a very expensive conversion is fine. I'll update the RFC later with a list of possible representations and their drawbacks.

Reply:
If we're `flit` is going to be just a string without (I assume) any compile-time arbitrary-precision arithmetic and the conversion happens at compile time anyway, I don't see any reason to have `flit` in the first place: just use strings and use the standard float parsing facilities (which aren't currently available at compile time, but should be).

Reply:
Can the standard float parsing facilities handle arbitrary-size floats? I assume you're referring to `f64::parse` and friends. For most use-cases what you suggest is totally fine, but I worry about restricting ourselves to the capacity of `f64`.

As I mentioned below, I think an opaque type (which could even just be a lang item!) might be a neater abstraction.

Reply:
For sure! Here's what I imagine if we go for the string-based alternative:
```rust
// mod core::ops? core::num is private iirc

#[lang = "float_lit"] // the compiler needs to know about this type, since it needs to
                      // be able to construct it from literals in source code.
struct flit(str); // sticking with flit for now, though since it's not a primitive kennytm
                  // will definitely want to go with FloatLit... which I would be ok with

impl flit {
    /// The literal, as it appears in source code.
    /// (Canonicalized to remove underscores.)
    ///
    /// Consider, e.g. `lit.verbatim().parse::<f64>().expect("bad literal")`
    const fn verbatim(&self) -> &str { .. }

    // the following are all *very* expensive str-manipulation fns
    const fn numer_bytes(&self) -> &[u8] { .. }
    const fn denom_bytes(&self) -> &[u8] { .. }

    const fn mantissa_bytes(&self, base: usize) -> &[u8] { .. }
    const fn exponent_bytes(&self, base: usize) -> &[u8] { .. }
}
```

Reply:
Who would benefit from `{numer,denom}_bytes` and how much? The only types that could use it that comes to mind are rationals, and it only saves them calling a function they'd probably have anyway (parsing decimal strings) -- and they may *not* otherwise have a way to interpret these u8 slices.

As for `mantissa_bytes`, `exponent_bytes`, I struggle to understand what these even do, let alone how they're useful for anything. The `(mantissa, exponent)` representation generally requires a periodic mantissa, since it's just a different way to write scientific notation. So, does it round? Then it's missing the target precision. But even the target precision is not enough for types with fixed-size exponent field, because overflowing the exponent range can impact rounding (e.g., if you have subnormals like IEEE 754).

What you *can* write is a parsing function for IEEE 754 floating point parametrized by the base, precision, and exponent range (that's the way the standard is written, even!) but that function is even more niche than the rational-based functions discussed before.

---

And yet again my most important contention remains unaddressed: what justifies putting these conversion functions into the standard library rather than letting those who need that conversion either do it themselves or import a third party library for it?